CMake - OpenMP Check fail with NVHPC

Hello, 

I was trying to build OpenBLAS 0.3.30 with NVHPC (25.5 and 25.7). However, cmake fails early with this message : 

> Multiple OpenMP runtime libraries detected. Mixed OpenMP runtime linkage is dangerous. You may pass -DOpenMP_LANG_LIB_NAMES and -DOpenMP_omp_LIBRARY to manually choose the OpenMP library

This comes from these lines in system.cmake , that were introduced between 0.3.29 and 0.3.30 : 
https://github.com/OpenMathLib/OpenBLAS/blob/d23680b81d5179ce6ae1ca5546303b81646ecac1/cmake/system.cmake#L415-L427

I dag a bit more and dumped some variables : 
```cmake
        message("OpenMP_C_FOUND       = ${OpenMP_C_FOUND}")
        message("OpenMP_Fortran_FOUND = ${OpenMP_Fortran_FOUND}")
        message("OpenMP_VERSION       = ${OpenMP_VERSION}")
        message("OpenMP_C_LIB_NAMES   = ${OpenMP_C_LIB_NAMES}")
        message("OpenMP_Fortran_LIB_NAMES = ${OpenMP_Fortran_LIB_NAMES}")
        message("OMP_C_LIB            = ${OMP_C_LIB}")
        message("OMP_Fortran_LIB      = ${OMP_Fortran_LIB}")
```
With GCC, I get 
```
OpenMP_C_FOUND       = TRUE
OpenMP_Fortran_FOUND = TRUE
OpenMP_VERSION       =
OpenMP_C_LIB_NAMES   = gomp;pthread
OpenMP_Fortran_LIB_NAMES = gomp;pthread
OMP_C_LIB            = /home/morvana/compilers/gcc/15.1.0/prefix/lib64/libgomp.so;/lib/x86_64-linux-gnu/libpthread.a
OMP_Fortran_LIB      = /home/morvana/compilers/gcc/15.1.0/prefix/lib64/libgomp.so;/lib/x86_64-linux-gnu/libpthread.a
```

But for NVHPC, I get this :
```
OpenMP_C_FOUND       = TRUE
OpenMP_Fortran_FOUND = TRUE
OpenMP_VERSION       =
OpenMP_C_LIB_NAMES   =
OpenMP_Fortran_LIB_NAMES =
OMP_C_LIB            = OMP_C_LIB-NOTFOUND
OMP_Fortran_LIB      = OMP_Fortran_LIB-NOTFOUND
```

So I have the feeling the FindOpenMP from CMake (https://github.com/Kitware/CMake/blob/master/Modules/FindOpenMP.cmake) is not working properly with NVidia OpenMP. Do you share my opinion, before I open an issue over there ?

I add these flags to the cmake line as workaround at the moment : 
```sh
            OPENBLAS_CMAKE_EXTRA_FLAGS="-DOpenMP_C_LIB_NAMES=nvomp -DOpenMP_Fortran_LIB_NAMES=nvomp"
            OPENBLAS_CMAKE_EXTRA_FLAGS+=" -DOpenMP_nvomp_LIBRARY=${NVHPC_ROOT}/compilers/lib/libnvomp.a"
```
Do you have a better one to propose ?

Best.

@ywwry66 - in case, NVHPC software is here: https://developer.nvidia.com/nvidia-hpc-sdk-255-downloads, and you don't need NVIDIA hardware to trigger the bug.
After installing it on an up to date Gentoo Linux box, and setting PATH etc, with current OpenBLAS from git

```
mkdir build && cd build
cmake -DCMAKE_C_COMPILER=nvc -DCMAKE_Fortran_COMPILER=nvfortran -DUSE_OPENMP=1 ..
```
I get
```
-- The ASM compiler identification is unknown
-- Found assembler: /opt/nvidia/hpc_sdk/Linux_x86_64/25.5/compilers/bin/nvc
CMake Warning at CMakeLists.txt:122 (message):
  CMake support is experimental.  It does not yet support all build options
  and may not produce the same Makefiles that OpenBLAS ships with.


-- GEMM multithread threshold set to 4.
-- Multi-threading enabled with 14 threads.
-- Running getarch
-- GETARCH results:
CORE=HASWELL
LIBCORE=haswell
NUM_CORES=14
HAVE_MMX=1
HAVE_SSE=1
HAVE_SSE2=1
HAVE_SSE3=1
HAVE_SSSE3=1
HAVE_SSE4_1=1
HAVE_SSE4_2=1
HAVE_AVX=1
HAVE_AVX2=1
HAVE_FMA3=1
MAKEFLAGS += -j 14

-- Compiling a 64-bit binary.
-- Found OpenMP_C: -mp (found version "5.1")
-- Found OpenMP: TRUE (found version "5.1") found components: C
-- Found OpenMP_Fortran: -mp (found version "5.1")
-- Found OpenMP: TRUE (found version "5.1") found components: Fortran
CMake Error at cmake/system.cmake:424 (message):
  Multiple OpenMP runtime libraries detected.  Mixed OpenMP runtime linkage
  is dangerous.  You may pass -DOpenMP_LANG_LIB_NAMES and
  -DOpenMP_omp_LIBRARY to manually choose the OpenMP library.
Call Stack (most recent call first):
  CMakeLists.txt:125 (include)


-- Configuring incomplete, errors occurred!
```
which demonstrates the problem with `FindOpenMP` as reported.

--- a/cmake/system.cmake
+++ b/cmake/system.cmake
@@ -418,10 +418,15 @@ if (USE_OPENMP)
   if (NOT NOFORTRAN)
     find_package(OpenMP COMPONENTS Fortran REQUIRED)
     # Avoid mixed OpenMP linkage
-    get_target_property(OMP_C_LIB OpenMP::OpenMP_C INTERFACE_LINK_LIBRARIES)
-    get_target_property(OMP_Fortran_LIB OpenMP::OpenMP_Fortran INTERFACE_LINK_LIBRARIES)
-    if (NOT OMP_C_LIB STREQUAL OMP_Fortran_LIB)
-      message(FATAL_ERROR "Multiple OpenMP runtime libraries detected. Mixed OpenMP runtime linkage is dangerous. You may pass -DOpenMP_LANG_LIB_NAMES and -DOpenMP_omp_LIBRARY to manually choose the OpenMP library.")
+    get_target_property(OMP_C_LIBS OpenMP::OpenMP_C INTERFACE_LINK_LIBRARIES)
+    get_target_property(OMP_F_LIBS OpenMP::OpenMP_Fortran INTERFACE_LINK_LIBRARIES)
+    if (NOT OMP_C_LIBS STREQUAL OMP_F_LIBS)
+      message(NOTICE
+        "OpenMP runtimes detected:\n"
+        "C=${OMP_C_LIBS}\n"
+        "Fortran=${OMP_F_LIBS}\n"
+        "Please check that they are the same OpenMP runtime implementation."
+      )
     endif()
   endif ()
 endif ()

--- a/cmake/system.cmake
+++ b/cmake/system.cmake
@@ -422,10 +422,11 @@ if (USE_OPENMP)
     get_target_property(OMP_F_LIBS OpenMP::OpenMP_Fortran INTERFACE_LINK_LIBRARIES)
     if (NOT OMP_C_LIBS STREQUAL OMP_F_LIBS)
       message(NOTICE
-        "OpenMP runtimes detected:\n"
+        "CMake detected different OpenMP libraries for C and Fortran:\n"
         "C=${OMP_C_LIBS}\n"
         "Fortran=${OMP_F_LIBS}\n"
-        "Please check that they are the same OpenMP runtime implementation."
+        "In case you encounter issues, please check that this is correct.\n"
+        "You may pass -DOpenMP_<lang>_LIB_NAMES and -DOpenMP_<libname>_LIBRARY to cmake to manually force the OpenMP library."
       )
     endif()
   endif ()

